airbyte-db/lib/src/main/java/io/airbyte/db/factory/DataSourceFactory.java

```diff
@@ -0,0 +1,251 @@
+        .withDatabase(database)
+        .withDriverClassName("org.postgresql.Driver")
+        .withHost(host)
+        .withPort(port)
+        .withPassword(password)
+        .withUsername(username)
+        .build();
+  }
+
+  /**
+   * Builder class used to configure and construct {@link DataSource} instances.
+   */
+  private static class DataSourceBuilder {
+
+    private static final Map<String, String> JDBC_URL_FORMATS = Map.of("org.postgresql.Driver", "jdbc:postgresql://%s:%d/%s",
+        "com.amazon.redshift.jdbc.Driver", "jdbc:redshift://%s:%d/%s",
+        "com.mysql.cj.jdbc.Driver", "jdbc:mysql://%s:%d/%s",
+        "com.microsoft.sqlserver.jdbc.SQLServerDriver", "jdbc:sqlserver://%s:%d/%s",
+        "oracle.jdbc.OracleDriver", "jdbc:oracle:thin:@%s:%d:%s",
+        "ru.yandex.clickhouse.ClickHouseDriver", "jdbc:ch://%s:%d/%s",
+        "org.mariadb.jdbc.Driver", "jdbc:mariadb://%s:%d/%s");
+
+    private Map<String, String> connectionProperties = Map.of();
+    private String database;
+    private String driverClassName = "org.postgresql.Driver";
+    private String host;
+    private String jdbcUrl;
+    private Integer maximumPoolSize = 5;
+    private Integer minimumPoolSize = 0;
+    private String password;
+    private Integer port = 5432;
+    private String username;
+
+    private DataSourceBuilder() {}
+
+    public DataSourceBuilder withConnectionProperties(final Map<String, String> connectionProperties) {
+      if (connectionProperties != null) {
+        this.connectionProperties = connectionProperties;
+      }
+      return this;
+    }
+
+    public DataSourceBuilder withDatabase(final String database) {
+      this.database = database;
+      return this;
+    }
+
+    public DataSourceBuilder withDriverClassName(final String driverClassName) {
+      this.driverClassName = driverClassName;
+      return this;
+    }
+
+    public DataSourceBuilder withHost(final String host) {
+      this.host = host;
+      return this;
+    }
+
+    public DataSourceBuilder withJdbcUrl(final String jdbcUrl) {
+      this.jdbcUrl = jdbcUrl;
+      return this;
+    }
+
+    public DataSourceBuilder withMaximumPoolSize(final Integer maximumPoolSize) {
+      if (maximumPoolSize != null) {
+        this.maximumPoolSize = maximumPoolSize;
+      }
+      return this;
+    }
+
+    public DataSourceBuilder withMinimumPoolSize(final Integer minimumPoolSize) {
+      if (minimumPoolSize != null) {
+        this.minimumPoolSize = minimumPoolSize;
+      }
+      return this;
+    }
+
+    public DataSourceBuilder withPassword(final String password) {
+      this.password = password;
+      return this;
+    }
+
+    public DataSourceBuilder withPort(final Integer port) {
+      if (port != null) {
+        this.port = port;
+      }
+      return this;
+    }
+
+    public DataSourceBuilder withUsername(final String username) {
+      this.username = username;
+      return this;
+    }
+
+    public DataSource build() {
```

Comment:
Oh, I see now what using lombok might be hard here.

Reply:
This builder may shift some as I do the actual refactor.  Once it is finalized, we can revisit it to see if it makes sense to leverage Lombok at that point.